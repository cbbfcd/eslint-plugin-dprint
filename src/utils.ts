import { readFileSync } from "node:fs";
import { extname } from "node:path";

const TS_EXTS = [".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs", ".mts", ".cts"];
const JSON_EXTS = [".json", ".jsonc", ".json5"];
const MARKDOWN_EXTS = [".md", ".markdown", ".mdown", ".mkd"];

export function detectLanguage(filename: string) {
  const ext = extname(filename);
  if (TS_EXTS.includes(ext)) { return "typescript"; }
  if (ext === ".toml") { return "toml"; }
  if (JSON_EXTS.includes(ext)) { return "json"; }
  if (MARKDOWN_EXTS.includes(ext)) { return "markdown"; }
  if (filename === "Dockerfile") { return "dockerfile"; }
}

export const hasNewlineOnly = (str: string) => ["\r", "\n", "\r\n"].includes(str);

export function pick<T extends Record<string, any>, K extends keyof T>(obj: T, pickKeys: K[] | readonly K[]) {
  const result = {} as Pick<T, K>;
  pickKeys.forEach((key) => {
    if (obj[key] !== undefined) { result[key] = obj[key]; }
  });
  return result;
}

export function omit<T extends object, K extends keyof T>(obj: T, omitKeys: K[] | readonly K[]) {
  const result = {} as Omit<T, K>;
  const keys = Object.keys(obj).filter((item) => !omitKeys.includes(item as K)) as Exclude<keyof T, K>[];
  keys.forEach((key) => {
    if (obj[key] !== undefined) { result[key] = obj[key]; }
  });
  return result;
}

export const readJson = (filename: string) => JSON.parse(readFileSync(filename, "utf8"));
