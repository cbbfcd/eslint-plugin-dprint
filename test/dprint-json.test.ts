import { join } from "node:path";

import { ESLint } from "eslint";
import { describe, expect, it } from "vitest";

import { dir } from "./utils";

const dirname = dir();

const eslint = new ESLint({
  baseConfig: {
    parserOptions: {
      ecmaVersion: "latest",
      ecmaFeatures: {
        jsx: true,
      },
      sourceType: "module",
    },
    plugins: ["dprint-integration"],
    rules: {
      "dprint-integration/dprint": [
        "error",
        { useDprintJson: join(dirname, "__fixtures__/dprint.json") },
      ],
    },
  },
  useEslintrc: false,
  ignore: false,
});

describe("dprint json config", () => {
  it("should resolve config", async () => {
    const result = await eslint.lintFiles(join(dirname, "./__fixtures__/dprint-json/**"));
    expect(result).toMatchSnapshot();
  });
});
