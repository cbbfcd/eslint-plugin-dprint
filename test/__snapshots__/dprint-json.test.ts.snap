// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dprint json config > should resolve config 1`] = `
[
  {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "C:\\\\Workspace\\\\eslint-plugin-dprint-integration\\\\test\\\\__fixtures__\\\\dprint-json\\\\a.ts",
    "fixableErrorCount": 1,
    "fixableWarningCount": 0,
    "messages": [
      {
        "column": 11,
        "endColumn": 15,
        "endLine": 1,
        "fix": {
          "range": [
            10,
            14,
          ],
          "text": "'1';
",
        },
        "line": 1,
        "message": "Replace \`\\"1\\";\` with \`'1';‚èé\`",
        "messageId": "replace",
        "nodeType": null,
        "ruleId": "dprint-integration/dprint",
        "severity": 2,
      },
    ],
    "source": "const a = \\"1\\";
",
    "suppressedMessages": [],
    "usedDeprecatedRules": [],
    "warningCount": 0,
  },
]
`;
